### DISCUSSION TITLE
Best environment to develop?

### DISCUSSION THREAD

**Discussion by olivier1980** [dxx-rebirth/dxx-rebirth#762] [Q&A] [answered]
Clion doesn't support scons, so I'm looking for other IDE's..what is the best way to do some debugging/building for a windows environment. Just using Visual Studio?

**Comment by vLKp** [comment, marked as answer]
I use a text editor, `scons`, `gdb`, and a terminal emulator in which to run `gdb`.  A window manager that allows you to easily raise and lower the game and your development windows may also be useful.  I do not develop on Windows, and only cross-compile for it.

**Reply by olivier1980** [reply]
Regarding cross-compiling, I don't think that's documented in the Install markdown yet?

**Comment by vLKp** [comment]
I find it surprising that a complete toolset for C and C++ development would be unable to run external build systems.  Invocation of `scons` is sufficiently similar to GNU make that I configure my editor's Make integration to run `scons` instead, and it works fine.  What problems did you encounter trying to use `scons`?

**Reply by olivier1980** [reply]
Well with my limited C++ experience I haven't tried a lot so far, it's just that the official documentation says it doesn't support it out of the box: https://intellij-support.jetbrains.com/hc/en-us/articles/207252755-Which-build-systems-are-supported-Do-you-plan-to-support-any-other-build-systems

**Comment by vLKp** [comment]
I see the issue now.  CLion is discussing this in the context of wanting to know how the compiler is invoked (for the benefit of its syntax highlighting and similar), not in the context of building the program (where you only want to produce a runnable program).  Yes, extracting that from an `SConstruct` file would be difficult, which is why I added support for the clang compile_commands feature, back before the SCons project wrote and released a similar option.  Set `compilation_database=compile_commands.json` to invoke the feature I wrote.  It seemed to work properly with clang tools back when I used it.  I've never used CLion, so I don't know if they can consume it, but I would hope so.

**Comment by vLKp** [comment]
> Regarding cross-compiling, I don't think that's documented in the Install markdown yet?

Correct.  Cross-compilation is a complex topic.  The short form is that you should prepare in your cross-toolchain the headers and libraries for the target platform, then pass CXX=`/path/to/cross-g++`, CHOST=cross-host-tuple, PKG_CONFIG=`/path/to/cross-pkg-config` so that `SConstruct` picks up the cross-toolchain.
